`@tagName()` fails with enum backed by signed int

### Zig Version

0.11.0-dev.3704+729a051e9

### Steps to Reproduce and Observed Behavior

it seems that using `@tagName()` with an i32 backed enum fails while it succeeds with a u32.

```zig
// /tmp/tmp.zig
pub const Ok = enum(u32) { a, b, c };
pub const Nok = enum(i32) { a, b, c };

test {
    _ = @tagName(Ok.a);
    _ = @tagName(Nok.a);
}
```

```sh
$ zig version
0.11.0-dev.3704+729a051e9

$ zig test /tmp/tmp.zig
/tmp/tmp.zig:6:9: error: no field with value '@intToEnum(tmp.Nok, 0)' in enum 'Nok'
    _ = @tagName(Nok.a);
        ^~~~~~~~~~~~~~~
/tmp/tmp.zig:2:17: note: declared here
pub const Nok = enum(i32) { a, b, c };
```


### Expected Behavior

test should pass whether the enum is backed by a signed int or an unsigned int

maybe a better reproduction: 
```zig
pub const Ok = enum(u32) { a, b, c };
pub const Nok = enum(i32) { a, b, c };

test {
    _ = @intToEnum(Ok, 0);
    _ = @intToEnum(Nok, 0);
}
```
```console
$ zig test /tmp/tmp.zig
/tmp/tmp.zig:10:9: error: enum 'tmp.Nok' has no tag with value '0'
    _ = @intToEnum(Nok, 0);
        ^~~~~~~~~~~~~~~~~~
/tmp/tmp.zig:3:17: note: enum declared here
pub const Nok = enum(i32) { a, b, c };
                ^~~~~~~~~~~~~~~~~~~~~
```

--- a/src/InternPool.zig
+++ b/src/InternPool.zig
@@ -448,7 +448,11 @@ pub const Key = union(enum) {
                     if (x >= self.names.len) return null;
                     return @intCast(u32, x);
                 },
-                .i64, .big_int => return null, // out of range
+                .i64 => |x| {
+                    if (x >= self.names.len or x < 0) return null;
+                    return @intCast(u32, x);
+                },
+                .big_int => return null, // out of range
                 .lazy_align, .lazy_size => unreachable,
             }
         }

--- a/src/InternPool.zig
+++ b/src/InternPool.zig
@@ -443,18 +443,12 @@ pub const Key = union(enum) {
                 return @intCast(u32, field_index);
             }
             // Auto-numbered enum. Convert `int_tag_val` to field index.
-            switch (ip.indexToKey(int_tag_val).int.storage) {
-                .u64 => |x| {
-                    if (x >= self.names.len) return null;
-                    return @intCast(u32, x);
-                },
-                .i64 => |x| {
-                    if (x >= self.names.len or x < 0) return null;
-                    return @intCast(u32, x);
-                },
-                .big_int => return null, // out of range
+            const field_index = switch (ip.indexToKey(int_tag_val).int.storage) {
+                inline .u64, .i64 => |x| std.math.cast(u32, x) orelse return null,
+                .big_int => |x| x.to(u32) catch return null,
                 .lazy_align, .lazy_size => unreachable,
-            }
+            };
+            return if (field_index < self.names.len) field_index else null;
         }
     };
 

--- a/test/behavior/enum.zig
+++ b/test/behavior/enum.zig
@@ -812,11 +812,6 @@ test "signed integer as enum tag" {
     try expect(@intFromEnum(SignedEnum.A2) == 1);
 }
 
-test "int to enum with signed tag type" {
-    const E = enum(i32) { a, b, c };
-    try expect(@intToEnum(E, 0) == .a);
-}
-
 test "enum with one member and custom tag type" {
     const E = enum(u2) {
         One,
@@ -1201,4 +1196,19 @@ test "enum tag from a local variable" {
     };
     const i = @enumFromInt(S.Int(u32), 0);
     try std.testing.expect(@intFromEnum(i) == 0);
+}
+
+test "auto-numbered enum with signed tag type" {
+    const E = enum(i32) { a, b };
+
+    try std.testing.expectEqual(@as(i32, 0), @intFromEnum(E.a));
+    try std.testing.expectEqual(@as(i32, 1), @intFromEnum(E.b));
+    try std.testing.expectEqual(E.a, @enumFromInt(E, 0));
+    try std.testing.expectEqual(E.b, @enumFromInt(E, 1));
+    try std.testing.expectEqual(E.a, @enumFromInt(E, @as(i32, 0)));
+    try std.testing.expectEqual(E.b, @enumFromInt(E, @as(i32, 1)));
+    try std.testing.expectEqual(E.a, @enumFromInt(E, @as(u32, 0)));
+    try std.testing.expectEqual(E.b, @enumFromInt(E, @as(u32, 1)));
+    try std.testing.expectEqualStrings("a", @tagName(E.a));
+    try std.testing.expectEqualStrings("b", @tagName(E.b));
 }

--- a/test/behavior/enum.zig
+++ b/test/behavior/enum.zig
@@ -812,6 +812,11 @@ test "signed integer as enum tag" {
     try expect(@intFromEnum(SignedEnum.A2) == 1);
 }
 
+test "int to enum with signed tag type" {
+    const E = enum(i32) { a, b, c };
+    try expect(@intToEnum(E, 0) == .a);
+}
+
 test "enum with one member and custom tag type" {
     const E = enum(u2) {
         One,

